// rabbitmq-stream-js-client/dist/super_stream_consumer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SuperStreamConsumer = void 0;
const subscribe_request_1 = require("./requests/subscribe_request");
class SuperStreamConsumer {
    constructor(handle, params) {
        this.handle = handle;
        this.consumers = new Map();
        this.consumerRef = params.consumerRef;
        this.locator = params.locator;
        this.partitions = params.partitions;
    }
    async start() {
        await Promise.all(this.partitions.map(async (p) => {
            const partitionConsumer = await this.locator.declareConsumer({ stream: p, consumerRef: this.consumerRef, offset: subscribe_request_1.Offset.first(), singleActive: true }, this.handle);
            this.consumers.set(p, partitionConsumer);
            return;
        }));
    }
    static async create(handle, params) {
        const superStreamConsumer = new SuperStreamConsumer(handle, params);
        await superStreamConsumer.start();
        return superStreamConsumer;
    }
    async close() {
        await Promise.all([...this.consumers.values()].map((c) => c.close()));
    }
}
exports.SuperStreamConsumer = SuperStreamConsumer;
